pallets/payment/src/lib.rs

```diff
@@ -0,0 +1,372 @@
+            who,
+            combined_fee,
+            withdraw_reason,
+            ExistenceRequirement::AllowDeath,
+        ) {
+            Ok(imbalance) => Ok(LiquidityInfo {
+                extra_fee_multiplier,
+                imbalance: Some(imbalance),
+            }),
+            Err(_) => Err(InvalidTransaction::Payment.into()),
+        }
+    }
+
+    fn correct_and_deposit_fee(
+        who: &T::AccountId,
+        _dispatch_info: &DispatchInfoOf<CallOf<T>>,
+        _post_info: &PostDispatchInfoOf<CallOf<T>>,
+        corrected_fee: Self::Balance,
+        tip: Self::Balance,
+        already_withdrawn: Self::LiquidityInfo,
+    ) -> Result<(), TransactionValidityError> {
+        let LiquidityInfo {
+            extra_fee_multiplier,
+            imbalance,
+        } = already_withdrawn;
+        if let Some(paid) = imbalance {
+            // Calculate the amont to refund: `refund` = `paid` - `corrected_fee` * `extra_fee_multiplier`
+            let refund_amount = paid
+                .peek()
+                .saturating_sub(extra_fee_multiplier.saturating_mul_int(corrected_fee));
+
+            // refund to the the account that paid the fees. If this fails, the account might
+            // have dropped below the existential balance. In that case we don't refund anything.
+            let refund_imbalance = T::Currency::deposit_into_existing(who, refund_amount)
+                .unwrap_or_else(|_| PositiveImbalanceOf::<T>::zero());
+            // merge the imbalance caused by paying the fees and refunding parts of it again.
+            let adjusted_paid = paid
+                .offset(refund_imbalance)
+                .same()
+                .map_err(|_| TransactionValidityError::Invalid(InvalidTransaction::Payment))?;
+            // Call someone else to handle the imbalance (fee and tip separately)
+            let (tip, fee) = adjusted_paid.split(tip);
+            T::FeeDepositor::on_unbalanceds(Some(fee).into_iter().chain(Some(tip)));
+        }
+        Ok(())
+    }
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+    use super::*;
+    use frame_system::pallet_prelude::*;
+
+    #[pallet::config]
+    pub trait Config:
+        frame_system::Config + pallet_transaction_payment::Config + pallet_authorship::Config
+    {
+        /// Filter for calls subbject for extra fees
+        type ExtraFeeCallFilter: Contains<CallOf<Self>>;
+
+        /// Type used to actually deposit collected fees
+        type FeeDepositor: OnUnbalanced<
+            <Self::Currency as Currency<Self::AccountId>>::NegativeImbalance,
+        >;
+
+        type Currency: LockableCurrency<Self::AccountId>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    #[pallet::without_storage_info]
+    pub struct Pallet<T>(_);
+
+    #[pallet::type_value]
+    pub fn FeeMultiplierOnEmpty() -> CustomFeeMultiplier {
+        CustomFeeMultiplier::saturating_from_integer(1)
```

Comment:
I would provide some constant for that multiplier, cause we re-use it many times